### PR TITLE
Propagated expansion errors farther through Driver implementation.

### PR DESCRIPTION
**Pull request by ceastlund**
Propagating the errors further allows expansion errors to be integrated better with linting and checking errors. Specifically, the expansion errors will be put first, so they can be reported before more general errors. Without this change, specific error messages from ppx implementation were being superseded by generic "Extension ... was not translated" errors in output sometimes.

### CODE REVIEW

**Review by panglesd — COMMENTED** [review]
I validate the propagation of errors through the driver to be able to decide the order in which they are added to the structure when we have all of them.

About the `loc_ghost` being true, I don't know the semantic of it and cannot find a ressource on it, so I cannot say why your approach is needed, and why it is better than changing `Location.Error.make` to force `loc_ghost = false`.

File: src/driver.ml [hunk 1]
```diff
@@ -376,10 +384,18 @@ module V2 = struct
   let register_transformation_using_ocaml_current_ast ?impl ?intf ?aliases name
       =
     let impl =
-      Option.map impl ~f:(Ppxlib_ast.Selected_ast.of_ocaml_mapper Structure)
+      Option.map impl ~f:(fun f ctx ast ->
+          ast
+          |> Ppxlib_ast.Selected_ast.to_ocaml Structure
+          |> f ctx
+          |> Ppxlib_ast.Selected_ast.of_ocaml Structure)
```
[the remark this answers is not in the document]
Comment (ceastlund) [reply src/driver.ml]: I have no idea. This may be an artifact of an intermediate development state. I've put it back.

**Comment by ceastlund** [comment]
As for errors and `loc_ghost` - there are checks in `ppxlib` to make sure that syntax is not duplicated in two places with the same non-ghost location. They are disabled by default, but we enable them inside janestreet. This helps us "play nice" with merlin, because it more reliably associates a location with the right syntax instead of some ppx artifact.

So I've changed the code to set `loc_ghost=true` for errors generated in addition to the "main" expanded syntax. It's not appropriate, I think, to always do this for errors, because sometimes the main/only result for expansion might be an error node, in which case it should not have a ghost location.

**Comment by pitag-ha** [comment]
Thanks for the detailed answer about `loc_ghost`, @ceastlund! I was wondering as well why you made the `loc_ghost=true` change, but now it makes total sense.

As a follow-up question just to make sure: also if the error's location itself is a ghost location, the user will see the right location in the error message, right? I'm asking because I've also seen the approach of just making the structure/signature item location a ghost location and leaving the children locations as were.

**Comment by ceastlund** [comment]
Ghost locations are still presented as the same location. They're just not treated as the primary AST node for that location, for things like merlin.

**Pull request merged**